python/dolfinx/fem/petsc.py

```diff
@@ -1118,3 +1122,141 @@ def interpolation_matrix(space0: _FunctionSpace, space1: _FunctionSpace) -> PETS
         Interpolation matrix.
     """
     return _interpolation_matrix(space0._cpp_object, space1._cpp_object)
+
+
+@functools.singledispatch
+def assign(
+    x0: typing.Union[npt.NDArray[np.floating], list[npt.NDArray[np.floating]]], x1: PETSc.Vec
```

Comment:
Shouldn't the typing also include `np.complexfloating` for complex values?

Reply:
From my reading of the numpy documentation this concept is expressed as `np.inexact`. https://numpy.org/doc/stable/reference/arrays.scalars.html#inexact-types

I suspect we're making a number of issues overtyping/undertyping/incorrectly typing throughout the codebase. This could be looked at in another PR.

Reply:
Clearly we should get this one agreed on so we can look elsewhere!